docs/auth-modes.md

```diff
@@ -0,0 +1,263 @@
+        ...
+        <connection-normalizer>
+            <required-attributes>
+                <attribute-list>
+                    ...
+                    <attr>username</attr>
+                    <attr>password</attr>
+                    ...
+                </attribute-list>
+            </required-attributes>
+        </connection-normalizer>
+    </connection-resolver>
+    ...
+</tdr>
+```
+
+### Multiple Authentication Modes
+
+User is prompted for which authentication option to use, then a set of fields appear, conditional on that option.  Depending on the option selected the user may or may not be prompted for credentials when reconnecting to the data source.
+
+Support authention options are: 
+```
+None
+Username
+Password
+UsernameAndPassword
+LDAP
+```
+
+```xml
+<!-- Connection Dialog -->
+<connection-dialog class='sample'>
+    <connection-config>
+        <authentication-mode value='ComboBoxIntegrated' />
+        <authentication-options>
+            <option name="None" value="auth-none" />
+            <option name="Username" value="username" />
+            <option name="UsernameAndPassword" value="username-password" default="true" />
+        </authentication-options>
+        ...
+    </connection-config>
+</connection-dialog>
+```
+
+Note: the ```value``` attribute value for all options are customizable by connector author, except None, which is required to be ```auth-none```.  These option values are the persisted value of the authentication attribute in a Tableau Workbook (twb) or Tableau Data Source (tds) file. 
+
+```xml
+<!-- Connection Resolver -->
+<tdr class='sample'>
+    <connection-resolver>
+        ...
+        <connection-normalizer>
+            <required-attributes>
+                <attribute-list>
+                    ...
+                    <attr>authentication</attr>
+                    <attr>username</attr>
+                    <attr>password</attr>
+                    ...
+                </attribute-list>
+            </required-attributes>
+        </connection-normalizer>
+    </connection-resolver>
+    ...
+</tdr>
+```
+
+```javascript
+// Connection Builder
+(function dsbuilder(attr)
+{
+    ...
+    var authAttrValue = attr[connectionHelper.attributeAuthentication];
+    if (authAttrValue == "auth-none")
+        // no-op
+    else if (authAttrValue == "username")
+        params["UID"] = attr[connectionHelper.attributeUsername];
+    else if (authAttrValue == "username-password")
+    {
+        params["UID"] = attr[connectionHelper.attributeUsername];
+        params["PWD"] = attr[connectionHelper.attributePassword];
+    }
+
+    ...
+})
+```
+
+## Considerations for 'hadoophive' and 'spark' base classes
+
+When using 'hadoophive' or 'spark' as a base class there are additional constraints and requirements in the Connection Dialog and Connection Resolver.
+
+The Connection Dialog authentication options ```option``` element's ```value``` attribute's value must match below.
```

Comment:
I'm having a hard time parsing this sentence -- is it possible to construct the sentence to say "this" must match "this"?

Reply:
I was having trouble with this section as well.  I think its a little better now.